aerogear/graphback
# feat: add database initialization strategies

## Motivation

Adding database initialization strategies is the first step towards building robust schema migrations in Graphback. There are currently 4 initialization strategies:

- `DropCreateDatabaseAlways` - this will always drop and recreate the database every time the application is run.
- `DropCreateDatabaseIfChanges` - if there are changes to the schema then drop the database and create it again. Otherwise don't do anything.
- `UpdateDatabaseIfChanges` - Only update the database if there are changes to the schema.
- `CreateDatabaseIfNotExists` - Create the database if it does not exist. **Not fully working. Knex creates the database on connect so the database always exists**.

These can be executed at runtime or through the CLI.

- `graphback db` - create the database if it doesn't exist.
- `graphback db:recreate` - drop and create the database.
- `graphback db:update` - update the database if there are changes.

## Tasks

- [x] Add `DropCreateDatabaseAlways` strategy
- [x] Add `DropCreateDatabaseIfChanges` strategy
- [x] Add `UpdateDatabaseIfChanges` strategy
- [ ] Add `CreateDatabaseIfNotExists` strategy (requires changes)

## Verification Steps [CLI]

1. Run `graphback db` to create database. This will only create a database if it doesn't exist.
2. Make changes to input schema (add a new field or type).
3. Run `graphback db:update`. The database schema should reflect the new field/type added to the schema.
4. Run `graphback db:recreate`. The database should be dropped and created again.

## Verification Steps [Runtime]

### DropCreateDatabaseAlways

1. Open `templates/apollo-runtime-ts/src/runtime.ts`.
2. Set initialization strategy to `DropCreateDatabaseAlways`:

```ts
const databaseInitializationStrategy = new DropCreateDatabaseAlways(config);
```

3. Run application: `npm run develop`.
4. The database should drop and recreate itself every time.

### DropCreateDatabaseIfChanges

1. Open `templates/apollo-runtime-ts/src/runtime.ts`.
2. Set initialization strategy to `DropCreateDatabaseIfChanges`:

```ts
const databaseInitializationStrategy = new DropCreateDatabaseIfChanges(config);
```

3. Make changes to the schema.
3. Run application: `npm run develop`.
4. The database should drop and recreate **only if** the schema has been changed.

### UpdateDatabaseIfChanges

1. Open `templates/apollo-runtime-ts/src/runtime.ts`.
2. Set initialization strategy to `UpdateDatabaseIfChanges`:

```ts
const databaseInitializationStrategy = new UpdateDatabaseIfChanges(config);
```

3. Make changes to the schema.
3. Run application: `npm run develop`.
4. The database schema should be updated with the latest changes.
